spring-session-core/src/main/java/org/springframework/session/config/annotation/web/http/PersistentSessionConfiguration.java

```diff
@@ -0,0 +1,104 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.session.config.annotation.web.http;
+
+import java.security.NoSuchAlgorithmException;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import javax.crypto.KeyGenerator;
+import javax.crypto.SecretKey;
+
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.ObjectProvider;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ApplicationContextAware;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Import;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.session.PersistentSessionPrincipalNameResolver;
+import org.springframework.session.PersistentSessionPrincipalRestorer;
+import org.springframework.session.PersistentSessionRepository;
+import org.springframework.session.SecurityContextAttributePersistentSessionRestorer;
+import org.springframework.session.SecurityContextAttributePrincipalNameResolver;
+import org.springframework.session.config.SessionRepositoryCustomizer;
+
+@Configuration(proxyBeanMethods = false)
+@Import(SpringHttpSessionConfiguration.class)
+public class PersistentSessionConfiguration implements ApplicationContextAware {
+
+	private ApplicationContext applicationContext;
+
+	private SecretKey secretKey;
+
+	@Bean
+	PersistentSessionRepository persistentSessionRepository(
+			ObjectProvider<PersistentSessionPrincipalNameResolver> principalNameResolverProvider,
+			ObjectProvider<PersistentSessionPrincipalRestorer> principalRestorerProvider,
+			ObjectProvider<SessionRepositoryCustomizer<PersistentSessionRepository>> sessionRepositoryCustomizers) {
+		PersistentSessionPrincipalNameResolver principalResolver = principalNameResolverProvider
+			.getIfAvailable(this::getDefaultPrincipalResolver);
+		PersistentSessionPrincipalRestorer principalRestorer = principalRestorerProvider
+			.getIfAvailable(this::getDefaultPrincipalRestorer);
+		PersistentSessionRepository repository = new PersistentSessionRepository(getSecretKey(), principalResolver,
+				principalRestorer);
+		sessionRepositoryCustomizers.orderedStream().forEach((customizer) -> customizer.customize(repository));
+		return repository;
+	}
+
+	@Autowired(required = false)
+	@Qualifier("springSessionPersistentSessionSecretKey")
+	public void setSecretKey(SecretKey secretKey) {
+		this.secretKey = secretKey;
```

Comment:
I think that using a qualifier is ok. However, you might consider providing a meta annotation so that users do not need to remember the qualifier name.

 I'm not sure but I think Qualifier only is used if there are multiple beans of the same type (you might double check that). If that is true, you might create a configuration object that could be provided that contains the secret key. That would also ensure that users don't accidentally using another SecretKey (which is a common type) with Spring Session.